swarm/animator.py

```diff
@@ -45,3 +59,306 @@ def animate(i):
+    if missing:
+        raise ValueError("plt.Axes does not have {}".format(",".join(missing)))
+
+
+def _kwargs_hook(**kwargs):
+    """
+    Our animation library uses hooks to allow once of configuration of an axes. These
+    include things like plotting the true y value or setting sensible x/y limits for the plot
+
+    This can be time consuming to do, so this provides an easy way to do so, allowing
+    one to call the various functions of ``plt.Axes`` as below. This is expected to be fit
+    into the existing machinery of the SwarmPlot sub/duck classes so you won't need to make
+    this call directly
+
+    Examples:
+        >>> _kwargs_hook(set_title="Goofy Experiment", set_ylabel="loss")
+        # will be used mostly with constructors like
+        >>> LineSwarm.standard(x, y, data, set_title="Goofy Experiment", set_ylabel="loss")
+
+    Raises:
+        ValueError: will attempt to validate the kwargs ahead of time, but also on execution
+        in case the function arg is incorrect
+
+    """
+    # get an early error if this is a problem
+    # rather than at animate time
+    _validate_ax_kwargs(**kwargs)
+
+    def inner_hook(ax):
+        for k, v in kwargs.items():
+            try:
+                func = getattr(ax, k)
+            except AttributeError:
+                raise ValueError(f"Used a method {k}")
+            else:
+                try:
+                    func(v)
+                except Exception as e:
+                    # we might have an issue that we're hitting a property
+                    # or have type mismatches
+                    raise ValueError("Unexpected failure") from e
+
+    return inner_hook
+
+
+def _apply_hook(ax: plt.Axes, hook: Union[Iterable[Callable], Callable, None]):
+    """Applies a list of hooks to the axes object"""
+    if not hook:
+        return
+    try:
+        for h in hook:
+            h(ax)
+    except TypeError:
+        # can't iterate, must be single callabe
+        hook(ax)
+
+
+@attr.s
+class LineSwarm(SwarmPlot):
+    """
+    This is the standard class
+    """
+
+    x = attr.ib(type=np.ndarray)  # shape of (N,)
+    data = attr.ib(type=np.ndarray)  # shape of (num_lines, timestep, N)
+    hook = attr.ib(type=List[Callable], factory=list)
+
+    @classmethod
+    def standard(cls, xd, yd, data, **kwargs):
+        """Plot against true yd"""
+        hook = [lambda ax: ax.plot(xd, yd, "."), _kwargs_hook(**kwargs)]
+        return cls(xd, data, hook)
+
+    @classmethod
+    def auto_range(cls, xd, data, **kwargs):
+        """In the abscence of a true value, make sensible choices"""
+        mx, mn = [data.max(), data.min()]
+
+        def set_lim(ax: plt.Axes):
+            ax.set_ylim(mn, mx)
+            ax.set_xlim(xd.min(), xd.max())
+
+        hook = [set_lim, _kwargs_hook(**kwargs)]
+        return cls(xd, data, hook)
+
+    @property
+    def num_frames(self) -> int:
+        return self.data.shape[1]
+
+    def init(self, ax: plt.Axes):
```

Comment:
So this name is a little misleading here. This `init` function represents the `plot_init` function used by matplotlib rather than the class init - `__init__`. I will change that to reduce confusion

Using `attrs` https://www.attrs.org/en/stable/examples.html will generate this `__init__` and goes a bit into my philosophy around class design.

Reply:
Ok, thanks! Yes, I think a different name would reduce confusion :-)